### PR TITLE
Daylight sensor

### PR DESCRIPTION
**Pull request by partypyro**
I've been doing some working trying to get the daylight sensor to work. I've got the output to work according to this function which I modeled after https://minecraft.gamepedia.com/File:Daylight_sensor_output_clear.png
daylightfunc

But I can't seem to get the sensor to directly output to a redstone wire attached, only to a adjacent block. Any suggestions from people familiar with the redstone simulator?

notworking
working

### CODE REVIEW

**Comment by tigerw** [comment]
Try adding the daylight sensors to here:

https://github.com/cuberite/cuberite/blob/1e477ad063f9d87f81fc18478897e9b090765820/src/Simulator/IncrementalRedstoneSimulator/ForEachSourceCallback.cpp#L94-L98

**Comment by tigerw** [comment]
Just noticed a typo

> Normally we don't ask solid blocks for power because they don't have any (***store***, dirt, etc.) 

mhm yes, the land upon which we stand is made of supermarket

**Comment by partypyro** [comment]
lol thank you I finally got it to work properly

fixes #4918

File: src/Simulator/IncrementalRedstoneSimulator/DaylightSensorHandler.h [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+
+
+
+
+namespace DaylightSensorHandler
+{
+	inline unsigned char GetPowerLevel(const cChunk & Chunk, const Vector3i Position)
+	{
+		double PI = 3.14159;
```
Comment (12xx12) [inline src/Simulator/IncrementalRedstoneSimulator/DaylightSensorHandler.h:12]: There should be a PI macro

Something like M_PI I think

File: src/Simulator/IncrementalRedstoneSimulator/DaylightSensorHandler.h [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+		UNUSED(a_QueryPosition);
+
+		// Daylight sensors output to all surrounding blocks
+		// Retrieve and return the cached power calculated by Update for performance
+		return IsLinked ? 0 : DataForChunk(a_Chunk).GetCachedPowerData(a_Position);
+	}
+
+	inline void Update(cChunk & a_Chunk, cChunk & CurrentlyTicking, Vector3i a_Position, BLOCKTYPE a_BlockType, NIBBLETYPE a_Meta, const PowerLevel Power)
+	{
+		// LOGD("Evaluating Darryl the daylight sensor (%d %d %d)", a_Position.x, a_Position.y, a_Position.z);
```
Comment (12xx12) [inline src/Simulator/IncrementalRedstoneSimulator/DaylightSensorHandler.h:38]: Some debug output left here

**Review by 12xx12 — COMMENTED** [review]
Some minor stuff

File: src/Simulator/IncrementalRedstoneSimulator/DaylightSensorHandler.h (hunk 2 above)
Comment (partypyro) [reply src/Simulator/IncrementalRedstoneSimulator/DaylightSensorHandler.h:38]: Thanks for the tips I'll fix those. The debug output was left there b/c it's in the other redstone handlers as well but I can remove it

Comment (12xx12) [reply src/Simulator/IncrementalRedstoneSimulator/DaylightSensorHandler.h:37]: Yeah - if it's everywhere we can keep the debug output

**Comment by partypyro** [comment]
Should be good to pull now!

File: src/Simulator/IncrementalRedstoneSimulator/DaylightSensorHandler.h [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+
+
+
+
+namespace DaylightSensorHandler
+{
+	inline unsigned char GetPowerLevel(const cChunk & Chunk, const Vector3i Position)
+	{
+		int TimeOfDay = Chunk.GetWorld()->GetTimeOfDay();
```
Comment (tigerw) [inline src/Simulator/IncrementalRedstoneSimulator/DaylightSensorHandler.h:12]: Daylight sensors also respond to stuff like current skylight level, current weather right? I know we have a cChunk::GetTimeAlteredLight but I'm unsure if it's applicable here.

File: src/Simulator/IncrementalRedstoneSimulator/DaylightSensorHandler.h [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+	inline PowerLevel GetPowerDeliveredToPosition(const cChunk & a_Chunk, Vector3i a_Position, BLOCKTYPE a_BlockType, Vector3i a_QueryPosition, BLOCKTYPE a_QueryBlockType, bool IsLinked)
+	{
+		UNUSED(a_Chunk);
+		UNUSED(a_BlockType);
+		UNUSED(a_QueryPosition);
+
+		// Daylight sensors output to all surrounding blocks
+		// Retrieve and return the cached power calculated by Update for performance
+		return IsLinked ? 0 : DataForChunk(a_Chunk).GetCachedPowerData(a_Position);
```
Comment (tigerw) [inline src/Simulator/IncrementalRedstoneSimulator/DaylightSensorHandler.h:32]: Sensors store their power level in the block metadata https://minecraft.gamepedia.com/Java_Edition_data_value/Pre-flattening#Daylight_Sensor so you can SetMeta/GetMeta directly.

File: src/Simulator/IncrementalRedstoneSimulator/RedstoneHandler.cpp [hunk 5]
```diff
@@ -35,46 +36,48 @@
...
-			case E_BLOCK_REDSTONE_WIRE:         return RedstoneWireHandler::Callback;       \
-			case E_BLOCK_TNT:                   return TNTHandler::Callback;                \
-			case E_BLOCK_TRAPPED_CHEST:         return TrappedChestHandler::Callback;       \
-			case E_BLOCK_TRIPWIRE_HOOK:         return TripwireHookHandler::Callback;       \
-			default:                                                                        \
+			case E_BLOCK_ACTIVATOR_RAIL:                                                                \
+			case E_BLOCK_DETECTOR_RAIL:                                                                 \
+			case E_BLOCK_DAYLIGHT_SENSOR:                   return DaylightSensorHandler::Callback;     \
+			case E_BLOCK_INVERTED_DAYLIGHT_SENSOR:          return DaylightSensorHandler::Callback;     \
+			case E_BLOCK_POWERED_RAIL:                      return PoweredRailHandler::Callback;        \
```
Comment (tigerw) [inline src/Simulator/IncrementalRedstoneSimulator/RedstoneHandler.cpp:43]: The return statements for ACTIVATOR_RAIL and DETECTOR_RAIL now fall through to DaylightSensorHandler, consider moving the daylight sensor lines to below POWERED_RAIL and deleting the extra `return DaylightSensorHandler`.

File: Server/Plugins/APIDump/APIDesc.lua [hunk 6]
```diff
@@ -14452,6 +14452,14 @@ end
 				{
 					Notes = "Damage from contact with a cactus block"
 				},
+				dtMagmaContact =
```
Comment (tigerw) [inline Server/Plugins/APIDump/APIDesc.lua:14455]: You may wish to `git rebase -i origin/master` and delete the extra commits :)

Comment (partypyro) [reply Server/Plugins/APIDump/APIDesc.lua:14455]: Will do!

**Comment by partypyro** [comment]
I rebased but something got messed up D: All the commits got duplicated... I'm not sure how to fix my local branch to get rid of these duplicates

**Pull request closed without merging**